src/Confluent.Kafka/Consumer.cs

```diff
@@ -28,28 +30,160 @@
+        ///     keeps track of whether or not assign has been called during
+        ///     invocation of a rebalance callback event.
         /// </summary>
-        public IDeserializer<TKey> KeyDeserializer { get; }
+        private bool assignCalled = false;
+
+        private readonly bool enableHeaderMarshaling = true;
+        private readonly bool enableTimestampMarshaling = true;
+        private readonly bool enableTopicNamesMarshaling = true;
+
+        private IDeserializer<TKey> KeyDeserializer { get; }
+        private IDeserializer<TValue> ValueDeserializer { get; }
+
+        private readonly SafeKafkaHandle kafkaHandle;
+
+        private static readonly byte[] EmptyBytes = new byte[0];
+
+
+        private Action<Error> errorDelegate = null;
+        private readonly Librdkafka.ErrorDelegate errorCallbackDelegate;
+        private void ErrorCallback(IntPtr rk, ErrorCode err, string reason, IntPtr opaque)
+        {
+            // Ensure registered handlers are never called as a side-effect of Dispose/Finalize (prevents deadlocks in common scenarios).
+            if (kafkaHandle.IsClosed) { return; }
+
+            if (errorDelegate != null)
+            {
+                errorDelegate(new Error(err, reason));
+            }
+        }
+
+        private Action<string> statsDelegate = null;
+        private readonly Librdkafka.StatsDelegate statsCallbackDelegate;
+        private int StatsCallback(IntPtr rk, IntPtr json, UIntPtr json_len, IntPtr opaque)
+        {
+            // Ensure registered handlers are never called as a side-effect of Dispose/Finalize (prevents deadlocks in common scenarios).
+            if (kafkaHandle.IsClosed) { return 0; }
+
+            if (statsDelegate != null)
+            {
+                statsDelegate(Util.Marshal.PtrToStringUTF8(json));
+            }
+
+            return 0; // instruct librdkafka to immediately free the json ptr.
+        }
+
+        private object loggerLockObj = new object();
+        private Action<LogMessage> logDelegate = null;
+        private readonly Librdkafka.LogDelegate logCallbackDelegate;
+        private void LogCallback(IntPtr rk, SyslogLevel level, string fac, string buf)
+        {
+            // Ensure registered handlers are never called as a side-effect of Dispose/Finalize (prevents deadlocks in common scenarios).
+            // Note: kafkaHandle can be null if the callback is during construction (in that case the delegate should be called).
+            if (kafkaHandle != null && kafkaHandle.IsClosed) { return; }
+
+            var name = Util.Marshal.PtrToStringUTF8(Librdkafka.name(rk));
+
+            lock (loggerLockObj)
+            {
+                if (logDelegate == null)
+                {
+                    // A stderr logger is used by default if none is specified.
+                    Loggers.ConsoleLogger(this, new LogMessage(name, level, fac, buf));
```

Comment:
why not set logDelegate by default (and when logDelegate is reset) instead of performing this check for each call?

Reply:
good point. this is a hangover from when this was an event, where you can have multiple handlers registered to the same event. it makes much less sense now.

Reply:
also changed on the producer.